File: study/examples/Algorithm/python/06_searching.py
"""
이분 탐색 (Binary Search)
Binary Search Algorithms

정렬된 데이터에서 O(log n) 시간에 검색하는 알고리즘입니다.
"""

from typing import List, Optional
import bisect


# =============================================================================
# 1. 기본 이분 탐색
# =============================================================================
def binary_search(arr: List[int], target: int) -> int:
    """
    정렬된 배열에서 target의 인덱스 찾기
    없으면 -1 반환
    시간복잡도: O(log n)
    """
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2  # 오버플로우 방지

        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def binary_search_recursive(arr: List[int], target: int, left: int, right: int) -> int:
    """이분 탐색 (재귀 버전)"""
    if left > right:
        return -1

    mid = left + (right - left) // 2

    if arr[mid] == target:
        return mid
    elif arr[mid] < target:
        return binary_search_recursive(arr, target, mid + 1, right)
    else:
        return binary_search_recursive(arr, target, left, mid - 1)


# =============================================================================
# 2. Lower Bound / Upper Bound
# =============================================================================
def lower_bound(arr: List[int], target: int) -> int:
    """
    target 이상인 첫 번째 요소의 인덱스
    모든 요소가 target보다 작으면 len(arr) 반환
    """
    left, right = 0, len(arr)

    while left < right:
        mid = left + (right - left) // 2
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid

    return left


def upper_bound(arr: List[int], target: int) -> int:
    """
    target 초과인 첫 번째 요소의 인덱스
    모든 요소가 target 이하이면 len(arr) 반환
    """
    left, right = 0, len(arr)

    while left < right:
        mid = left + (right - left) // 2
        if arr[mid] <= target:
            left = mid + 1
        else:
            right = mid

    return left


def count_occurrences(arr: List[int], target: int) -> int:
    """특정 값의 등장 횟수"""
    return upper_bound(arr, target) - lower_bound(arr, target)


# =============================================================================
# 3. 첫 번째/마지막 위치 찾기
# =============================================================================
def find_first_position(arr: List[int], target: int) -> int:
    """target이 처음 등장하는 인덱스 (없으면 -1)"""
    idx = lower_bound(arr, target)
    if idx < len(arr) and arr[idx] == target:
        return idx
    return -1


def find_last_position(arr: List[int], target: int) -> int:
    """target이 마지막으로 등장하는 인덱스 (없으면 -1)"""
    idx = upper_bound(arr, target) - 1
    if idx >= 0 and arr[idx] == target:
        return idx
    return -1


# =============================================================================
# 4. 회전 정렬 배열 검색
# =============================================================================
def search_rotated(arr: List[int], target: int) -> int:
    """
    회전된 정렬 배열에서 검색
    예: [4, 5, 6, 7, 0, 1, 2] - 원래 [0,1,2,4,5,6,7]을 회전
    시간복잡도: O(log n)
    """
    if not arr:
        return -1

    left, right = 0, len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2

        if arr[mid] == target:
            return mid

        # 왼쪽 절반이 정렬되어 있는 경우
        if arr[left] <= arr[mid]:
            if arr[left] <= target < arr[mid]:
                right = mid - 1
            else:
                left = mid + 1
        # 오른쪽 절반이 정렬되어 있는 경우
        else:
            if arr[mid] < target <= arr[right]:
                left = mid + 1
            else:
                right = mid - 1

    return -1


# =============================================================================
# 5. 최솟값 찾기 (회전 배열)
# =============================================================================
def find_minimum_rotated(arr: List[int]) -> int:
    """회전 정렬 배열의 최솟값 찾기"""
    left, right = 0, len(arr) - 1

    while left < right:
        mid = left + (right - left) // 2

        if arr[mid] > arr[right]:
            left = mid + 1
        else:
            right = mid

    return arr[left]


# =============================================================================
# 6. 제곱근 구하기 (정수)
# =============================================================================
def integer_sqrt(n: int) -> int:
    """
    n의 정수 제곱근 (버림)
    예: sqrt(8) = 2
    """
    if n < 0:
        raise ValueError("음수의 제곱근은 정의되지 않습니다")
    if n == 0:
        return 0

    left, right = 1, n

    while left <= right:
        mid = left + (right - left) // 2
        square = mid * mid

        if square == n:
            return mid
        elif square < n:
            left = mid + 1
        else:
            right = mid - 1

    return right  # floor(sqrt(n))


# =============================================================================
# 7. 매개변수 탐색 (Parametric Search)
# =============================================================================
def can_split(arr: List[int], m: int, max_sum: int) -> bool:
    """
    배열을 m개 이하의 그룹으로 나눌 때
    각 그룹 합이 max_sum 이하인지 확인
    """
    count = 1
    current_sum = 0

    for num in arr:
        if current_sum + num > max_sum:
            count += 1
            current_sum = num
            if count > m:
                return False
        else:
            current_sum += num

    return True


def split_array_min_largest_sum(arr: List[int], m: int) -> int:
    """
    배열을 m개의 연속 부분 배열로 나눌 때
    각 부분 배열 합의 최댓값을 최소화
    시간복잡도: O(n log(sum))
    """
    left = max(arr)      # 최소 가능한 값: 가장 큰 요소
    right = sum(arr)     # 최대 가능한 값: 전체 합

    while left < right:
        mid = left + (right - left) // 2
        if can_split(arr, m, mid):
            right = mid
        else:
            left = mid + 1

    return left


# =============================================================================
# 8. Peak Element 찾기
# =============================================================================
def find_peak_element(arr: List[int]) -> int:
    """
    배열에서 peak element의 인덱스 찾기
    peak: arr[i] > arr[i-1] and arr[i] > arr[i+1]
    시간복잡도: O(log n)
    """
    left, right = 0, len(arr) - 1

    while left < right:
        mid = left + (right - left) // 2

        if arr[mid] > arr[mid + 1]:
            right = mid
        else:
            left = mid + 1

    return left


# =============================================================================
# 테스트
# =============================================================================
def main():
    print("=" * 60)
    print("이분 탐색 (Binary Search) 예제")
    print("=" * 60)

    # 1. 기본 이분 탐색
    print("\n[1] 기본 이분 탐색")
    arr = [1, 3, 5, 7, 9, 11, 13, 15]
    target = 7
    result = binary_search(arr, target)
    print(f"    배열: {arr}")
    print(f"    {target} 찾기 -> 인덱스: {result}")

    # 2. Lower/Upper Bound
    print("\n[2] Lower Bound / Upper Bound")
    arr = [1, 2, 2, 2, 3, 4, 5]
    target = 2
    lb = lower_bound(arr, target)
    ub = upper_bound(arr, target)
    count = count_occurrences(arr, target)
    print(f"    배열: {arr}")
    print(f"    target={target}")
    print(f"    lower_bound: {lb}, upper_bound: {ub}")
    print(f"    등장 횟수: {count}")

    # bisect 모듈과 비교
    print(f"    (bisect_left: {bisect.bisect_left(arr, target)}, "
          f"bisect_right: {bisect.bisect_right(arr, target)})")

    # 3. 첫 번째/마지막 위치
    print("\n[3] 첫 번째/마지막 위치")
    arr = [5, 7, 7, 8, 8, 8, 10]
    target = 8
    first = find_first_position(arr, target)
    last = find_last_position(arr, target)
    print(f"    배열: {arr}")
    print(f"    {target}의 첫 위치: {first}, 마지막 위치: {last}")

    # 4. 회전 배열 검색
    print("\n[4] 회전 정렬 배열 검색")
    arr = [4, 5, 6, 7, 0, 1, 2]
    target = 0
    result = search_rotated(arr, target)
    print(f"    회전 배열: {arr}")
    print(f"    {target} 찾기 -> 인덱스: {result}")

    # 5. 회전 배열 최솟값
    print("\n[5] 회전 배열 최솟값")
    arr = [4, 5, 6, 7, 0, 1, 2]
    result = find_minimum_rotated(arr)
    print(f"    회전 배열: {arr}")
    print(f"    최솟값: {result}")

    # 6. 제곱근
    print("\n[6] 정수 제곱근")
    for n in [4, 8, 16, 17, 100]:
        result = integer_sqrt(n)
        print(f"    sqrt({n}) = {result}")

    # 7. 매개변수 탐색
    print("\n[7] 매개변수 탐색 (배열 분할)")
    arr = [7, 2, 5, 10, 8]
    m = 2
    result = split_array_min_largest_sum(arr, m)
    print(f"    배열: {arr}, 분할 수: {m}")
    print(f"    최소 최대합: {result}")

    # 8. Peak Element
    print("\n[8] Peak Element 찾기")
    arr = [1, 2, 1, 3, 5, 6, 4]
    result = find_peak_element(arr)
    print(f"    배열: {arr}")
    print(f"    peak 인덱스: {result} (값: {arr[result]})")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: study/examples/Algorithm/cpp/06_searching.cpp
/*
 * 탐색 알고리즘 (Searching Algorithms)
 * Binary Search, Lower/Upper Bound, Parametric Search
 *
 * 효율적인 탐색 기법들입니다.
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

// =============================================================================
// 1. 이분 탐색 기본
// =============================================================================

int binarySearch(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid - 1;
    }

    return -1;
}

// 재귀 버전
int binarySearchRecursive(const vector<int>& arr, int target, int left, int right) {
    if (left > right) return -1;

    int mid = left + (right - left) / 2;

    if (arr[mid] == target)
        return mid;
    else if (arr[mid] < target)
        return binarySearchRecursive(arr, target, mid + 1, right);
    else
        return binarySearchRecursive(arr, target, left, mid - 1);
}

// =============================================================================
// 2. Lower Bound / Upper Bound
// =============================================================================

// target 이상인 첫 번째 위치
int lowerBound(const vector<int>& arr, int target) {
    int left = 0, right = arr.size();

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid;
    }

    return left;
}

// target 초과인 첫 번째 위치
int upperBound(const vector<int>& arr, int target) {
    int left = 0, right = arr.size();

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] <= target)
            left = mid + 1;
        else
            right = mid;
    }

    return left;
}

// target의 개수
int countOccurrences(const vector<int>& arr, int target) {
    return upperBound(arr, target) - lowerBound(arr, target);
}

// =============================================================================
// 3. 회전 정렬 배열 탐색
// =============================================================================

int searchRotated(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
            return mid;

        // 왼쪽이 정렬됨
        if (arr[left] <= arr[mid]) {
            if (arr[left] <= target && target < arr[mid])
                right = mid - 1;
            else
                left = mid + 1;
        }
        // 오른쪽이 정렬됨
        else {
            if (arr[mid] < target && target <= arr[right])
                left = mid + 1;
            else
                right = mid - 1;
        }
    }

    return -1;
}

// 회전 배열의 최솟값
int findMin(const vector<int>& arr) {
    int left = 0, right = arr.size() - 1;

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] > arr[right])
            left = mid + 1;
        else
            right = mid;
    }

    return arr[left];
}

// =============================================================================
// 4. 파라메트릭 서치
// =============================================================================

// 나무 자르기 (최대 높이)
long long cutWood(const vector<int>& trees, int h) {
    long long total = 0;
    for (int tree : trees) {
        if (tree > h) {
            total += tree - h;
        }
    }
    return total;
}

int maxCuttingHeight(const vector<int>& trees, long long need) {
    int left = 0;
    int right = *max_element(trees.begin(), trees.end());
    int result = 0;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (cutWood(trees, mid) >= need) {
            result = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return result;
}

// 배송 최소 용량
bool canShip(const vector<int>& weights, int capacity, int days) {
    int currentWeight = 0;
    int dayCount = 1;

    for (int w : weights) {
        if (w > capacity) return false;
        if (currentWeight + w > capacity) {
            dayCount++;
            currentWeight = w;
        } else {
            currentWeight += w;
        }
    }

    return dayCount <= days;
}

int shipWithinDays(const vector<int>& weights, int days) {
    int left = *max_element(weights.begin(), weights.end());
    int right = 0;
    for (int w : weights) right += w;

    while (left < right) {
        int mid = left + (right - left) / 2;
        if (canShip(weights, mid, days))
            right = mid;
        else
            left = mid + 1;
    }

    return left;
}

// =============================================================================
// 5. 실수 이분 탐색
// =============================================================================

// 제곱근 (소수점까지)
double sqrtBinary(double x, double precision = 1e-10) {
    if (x < 0) return -1;
    if (x < 1) {
        double lo = x, hi = 1;
        while (hi - lo > precision) {
            double mid = (lo + hi) / 2;
            if (mid * mid < x)
                lo = mid;
            else
                hi = mid;
        }
        return (lo + hi) / 2;
    }

    double lo = 1, hi = x;
    while (hi - lo > precision) {
        double mid = (lo + hi) / 2;
        if (mid * mid < x)
            lo = mid;
        else
            hi = mid;
    }
    return (lo + hi) / 2;
}

// =============================================================================
// 6. 2D 배열 탐색
// =============================================================================

// 행과 열이 정렬된 2D 배열 탐색
bool searchMatrix(const vector<vector<int>>& matrix, int target) {
    if (matrix.empty()) return false;

    int rows = matrix.size();
    int cols = matrix[0].size();
    int row = 0, col = cols - 1;

    while (row < rows && col >= 0) {
        if (matrix[row][col] == target)
            return true;
        else if (matrix[row][col] > target)
            col--;
        else
            row++;
    }

    return false;
}

// =============================================================================
// 테스트
// =============================================================================

int main() {
    cout << "============================================================" << endl;
    cout << "탐색 알고리즘 예제" << endl;
    cout << "============================================================" << endl;

    // 1. 이분 탐색
    cout << "\n[1] 이분 탐색" << endl;
    vector<int> arr = {1, 3, 5, 7, 9, 11, 13, 15};
    cout << "    배열: [1,3,5,7,9,11,13,15]" << endl;
    cout << "    7의 위치: " << binarySearch(arr, 7) << endl;
    cout << "    6의 위치: " << binarySearch(arr, 6) << endl;

    // 2. Lower/Upper Bound
    cout << "\n[2] Lower/Upper Bound" << endl;
    vector<int> arr2 = {1, 2, 2, 2, 3, 4, 5};
    cout << "    배열: [1,2,2,2,3,4,5]" << endl;
    cout << "    lower_bound(2): " << lowerBound(arr2, 2) << endl;
    cout << "    upper_bound(2): " << upperBound(arr2, 2) << endl;
    cout << "    2의 개수: " << countOccurrences(arr2, 2) << endl;

    // 3. 회전 정렬 배열
    cout << "\n[3] 회전 정렬 배열" << endl;
    vector<int> rotated = {4, 5, 6, 7, 0, 1, 2};
    cout << "    배열: [4,5,6,7,0,1,2]" << endl;
    cout << "    0의 위치: " << searchRotated(rotated, 0) << endl;
    cout << "    최솟값: " << findMin(rotated) << endl;

    // 4. 파라메트릭 서치
    cout << "\n[4] 파라메트릭 서치" << endl;
    vector<int> trees = {20, 15, 10, 17};
    cout << "    나무: [20,15,10,17], 필요량: 7" << endl;
    cout << "    최대 절단 높이: " << maxCuttingHeight(trees, 7) << endl;

    vector<int> weights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    cout << "    물건: [1-10], 5일 배송" << endl;
    cout << "    최소 용량: " << shipWithinDays(weights, 5) << endl;

    // 5. 실수 이분 탐색
    cout << "\n[5] 실수 이분 탐색" << endl;
    cout << "    sqrt(2) = " << sqrtBinary(2) << endl;
    cout << "    sqrt(10) = " << sqrtBinary(10) << endl;

    // 6. 2D 배열 탐색
    cout << "\n[6] 2D 배열 탐색" << endl;
    vector<vector<int>> matrix = {
        {1, 4, 7, 11},
        {2, 5, 8, 12},
        {3, 6, 9, 16},
        {10, 13, 14, 17}
    };
    cout << "    5 찾기: " << (searchMatrix(matrix, 5) ? "있음" : "없음") << endl;
    cout << "    15 찾기: " << (searchMatrix(matrix, 15) ? "있음" : "없음") << endl;

    // 7. 복잡도 요약
    cout << "\n[7] 복잡도 요약" << endl;
    cout << "    | 알고리즘          | 시간복잡도 |" << endl;
    cout << "    |-------------------|------------|" << endl;
    cout << "    | 선형 탐색         | O(n)       |" << endl;
    cout << "    | 이분 탐색         | O(log n)   |" << endl;
    cout << "    | 파라메트릭 서치   | O(log M × f(n)) |" << endl;
    cout << "    | 2D 행렬 탐색      | O(m + n)   |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}
